applications/FluidDynamicsApplication/custom_elements/alternative_qs_vms_dem_coupled.cpp

```diff
@@ -0,0 +1,685 @@
+{
+    QSVMS<TElementData>::Calculate(rVariable, rOutput, rCurrentProcessInfo);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+/**
+ * @see AlternativeQSVMSDEMCoupled::EquationIdVector
+ **/
+template < class TElementData >
+void AlternativeQSVMSDEMCoupled<TElementData>::EquationIdVector(
+    EquationIdVectorType& rResult,
+    const ProcessInfo& rCurrentProcessInfo) const
+{
+    QSVMS<TElementData>::EquationIdVector(rResult, rCurrentProcessInfo);
+}
+
+template< class TElementData >
+std::string AlternativeQSVMSDEMCoupled<TElementData>::Info() const
+{
+    std::stringstream buffer;
+    buffer << "AlternativeQSVMSDEMCoupled #" << this->Id();
+    return buffer.str();
+}
+
+
+template< class TElementData >
+void AlternativeQSVMSDEMCoupled<TElementData>::PrintInfo(std::ostream& rOStream) const
+{
+    rOStream << "AlternativeQSVMSDEMCoupled" << Dim << "D";
+}
+
+template<class TElementData>
+void AlternativeQSVMSDEMCoupled<TElementData>::CalculateRightHandSide(
+    VectorType& rRightHandSideVector,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+        TElementData data;
+        data.Initialize(*this, rCurrentProcessInfo);
+
+        // Calculate this element RHS contribution
+        QSVMS<TElementData>::CalculateRightHandSide(rRightHandSideVector, rCurrentProcessInfo);
+}
+
+template< class TElementData >
+void AlternativeQSVMSDEMCoupled<TElementData>::AlgebraicMomentumResidual(
+    const TElementData& rData,
+    const array_1d<double,3> &rConvectionVelocity,
+    array_1d<double,3>& rResidual) const
+{
+    const GeometryType rGeom = this->GetGeometry();
+
+    Vector convection; // u * grad(N)
+    this->ConvectionOperator(convection,rConvectionVelocity,rData.DN_DX);
+
+    const double density = this->GetAtCoordinate(rData.Density,rData.N);
+    const double viscosity = this->GetAtCoordinate(rData.DynamicViscosity, rData.N);
+    const auto& fluid_fraction = this->GetAtCoordinate(rData.FluidFraction, rData.N);
+    BoundedMatrix<double,Dim,Dim> permeability = this->GetAtCoordinate(rData.Permeability, rData.N);
+    BoundedMatrix<double,Dim,Dim> sigma = ZeroMatrix(Dim, Dim);
+    const auto& r_body_forces = rData.BodyForce;
+    const auto& r_velocities = rData.Velocity;
+    const auto& r_pressures = rData.Pressure;
+    const auto& fluid_fraction_gradient = this->GetAtCoordinate(rData.FluidFractionGradient, rData.N);
+
+    double det_permeability = MathUtils<double>::Det(permeability);
+    MathUtils<double>::InvertMatrix(permeability, sigma, det_permeability, -1.0);
+    sigma *= viscosity;
+
+    for (unsigned int i = 0; i < NumNodes; i++) {
+            const array_1d<double,3>& r_acceleration = rGeom[i].FastGetSolutionStepValue(ACCELERATION);
+            Vector sigma_U = ZeroVector(Dim);
+            BoundedMatrix<double,Dim,Dim> sym_gradient_u = ZeroMatrix(Dim, Dim);
+            Vector grad_alpha_sym_grad_u = ZeroVector(Dim);
```

Comment:
For the sake of performance it would be better to allocate all these arrays outside the loop. Otherwise you're creating and destroying for each `i`.